src/crypto/openssl/openssl_crypto_accel.cc

```diff
@@ -13,8 +13,68 @@
  */
 
 #include "crypto/openssl/openssl_crypto_accel.h"
-#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/engine.h>
+#include "common/debug.h"
 
+// -----------------------------------------------------------------------------
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_crypto
+#undef dout_prefix
+#define dout_prefix _prefix(_dout)
+
+static ostream&
+_prefix(std::ostream* _dout)
+{
+  return *_dout << "OpensslCryptoAccel: ";
+}
+// -----------------------------------------------------------------------------
+
+#define EVP_SUCCESS 1
+#define AES_ENCRYPT 1
+#define AES_DECRYPT 0
+
+bool evp_transform(unsigned char* out, const unsigned char* in, size_t size,
+                   const unsigned char* iv,
+                   const unsigned char* key,
+                   ENGINE* engine,
+                   const EVP_CIPHER* const type,
+                   const bool encrypt)
+{
+  using pctx_t = std::unique_ptr<EVP_CIPHER_CTX, decltype(&::EVP_CIPHER_CTX_free)>;
+  pctx_t pctx{ EVP_CIPHER_CTX_new(), EVP_CIPHER_CTX_free };
```

Comment:
can we reuse the `EVP_CIPHER_CTX` when encrypt/decrypt the memory chunk  and only create it in the constructor of `OpenSSLCryptoAccel`, and destroy it in the destructor?

Reply:
@tchaikov , it would be nice to avoid create & destroy the `EVP_CIPHER_CTX` when every encrypt/decrypt operation, but this may cause conflicts for there would be only one instance of crypto plugin between several rgw threads. every thread should have its own evp cipher status, so  the process would crash.
as far as I can see, the reason why there is one instance of crypto plugin is that all rgw threads share one same ceph context (`g_ceph_context`), so when every time calles `cct->get_plugin_registry` , it gets the same crypto plugin, and finally the same `EVP_CIPHER_CTX` (if just modified as above).  do u have any advice about this?

Reply:
i see. it's indeed embarrassing -- the curse of sharing.

@cbodley any idea?

Reply:
there's a distinction between the `CryptoPlugin`, which is stored in the global plugin registry, and the `CryptoAccel` which is instantiated by `CryptoPlugin::factory()` for each request. i think it's safe to assume that a single instance of `OpenSSLCryptoAccel` is thread-safe, so you should at least be able to share this `EVP_CIPHER_CTX` over multiple calls to `cbc_encrypt/decrypt()`

Reply:
i am still kind of confused here, just let me try to describe the process

when every rgw op which need encrypt/decrypt, it would call `get_crypto_accel` in `rgw_crypt.cc`, let's start from here.  main work of this func are as following:

1. to get a `CryptoPlugin`(OpenSSLCryptoPlugin here)
```
PluginRegistry *reg = cct->get_plugin_registry();
CryptoPlugin *factory = dynamic_cast<CryptoPlugin*>(reg->get_with_load("crypto", crypto_accel_type));
```
because cct is shared between rgw threads, and once a `CryptoPlugin` is load, it would  be stored globally, so all threads would share a same `CryptoPlugin` .
 
2. to get a `CryptoAccelRef` (OpenSSLCryptoAccel ref here)
it calls `factory->factory(&ca_impl, &ss);`
i think this would get a same `CryptoAccelRef` between threads for there is only one `OpenSSLCryptoPlugin`, and then `cryptoaccel`:
```
class OpenSSLCryptoPlugin : public CryptoPlugin {

  CryptoAccelRef cryptoaccel;
  int count = 0;
public:
  explicit OpenSSLCryptoPlugin(CephContext* cct) : CryptoPlugin(cct)
  {}
  int factory(CryptoAccelRef *cs, ostream *ss) override {
    if (cryptoaccel == nullptr)
      cryptoaccel = CryptoAccelRef(new OpenSSLCryptoAccel);

    *cs = cryptoaccel;
    return 0;
  }
};
```
so if making the `EVP_CIPHER_CTX` a member of `OpenSSLCryptoAccel`,  all threads would share a evp ctx with own evp state, then the process would crash. 

besides, i added log in `get_crypto_acceltest`  and tested these using cosbench  ,then got following results: 

2 threads doing 10 ops **with not adding** the `EVP_CIPHER_CTX`  to `OpenSSLCryptoAccel` , the test succeeded  with all ops getting a same  `OpenSSLCryptoAccel`

2 threads doing 10 ops **with adding** the `EVP_CIPHER_CTX`  to `OpenSSLCryptoAccel`, the process crashed

1 thread doing 10 ops **with adding** the `EVP_CIPHER_CTX`  to `OpenSSLCryptoAccel`, the test succeeded  with all ops getting a same  `OpenSSLCryptoAccel`

@cbodley @tchaikov  anything i did wrong above?

Reply:
@egggHang i concur with your observations.

> the `CryptoAccel` which is instantiated by `CryptoPlugin::factory()` for each request.

i think the crux of the problem is that the `CryptoAccel` instance created by its factory is cached by, for instance, `OpenSSLCryptoPlugin::factory()`, so only the first `CryptoAccel` instance is created, and all succeeding calls will return the same instance. as long as the same plugin is used for creating `CryptoAccelRef`s.